core-rust/state-manager/src/staging/cache.rs

```diff
@@ -93,20 +90,45 @@ use radix_engine_stores::hash_tree::tree_store::{
 use sbor::rust::collections::HashMap;
 use slotmap::SecondaryMap;
 
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, PartialOrd)]
+struct TransactionPlacement {
+    parent_transaction_root: TransactionTreeHash,
+    legacy_payload_hash: LegacyLedgerPayloadHash,
+}
+
+impl TransactionPlacement {
+    fn new(
+        parent_transaction_root: &TransactionTreeHash,
+        legacy_payload_hash: &LegacyLedgerPayloadHash,
+    ) -> Self {
+        Self {
+            parent_transaction_root: *parent_transaction_root,
+            legacy_payload_hash: *legacy_payload_hash,
+        }
+    }
+}
+
+struct InternalTransactionIds {
+    transaction_placement: TransactionPlacement,
+    committed_transaction_root: Option<TransactionTreeHash>,
+}
+
 pub struct ExecutionCache {
     stage_tree: StageTree<ProcessedTransactionReceipt, ImmutableStore>,
-    root_accumulator_hash: AccumulatorHash,
-    accumulator_hash_to_key: HashMap<AccumulatorHash, DerivedStageKey>,
-    key_to_accumulator_hash: SecondaryMap<DerivedStageKey, AccumulatorHash>,
+    base_transaction_root: TransactionTreeHash,
+    transaction_placement_to_key: HashMap<TransactionPlacement, DerivedStageKey>,
+    transaction_root_to_key: HashMap<TransactionTreeHash, DerivedStageKey>,
```

Comment:
@malexandru-rdx This bit may be interesting to you: _(how I got rid of accu hash in the staging tree, while avoiding calculation of a transaction merkle tree update before cache-hit)_

A transaction may be unambiguously identified by 2 different business keys here:

- by a "transaction placement":
  - A "transaction placement" of transaction X is a tuple `{X's parent's transaction root, X's payload hash}`. It can be produced for any "candidate" transaction (i.e. even for ones that would be rejected).

- by a transaction root:
   - Yup, just a regular transaction tree root (i.e. our replacement of accumulator hash). There is a gotcha though: transaction root comes from transaction's `LedgerHashes`, and we do not compute `LedgerHashes` for transactions that are rejected (technically we could produce just the transaction root alone for them, but in our current impl we don't). Yet, we want to reference rejected transactions in our cache too!
   - The transaction root is perfect for referencing every possible `{parent of a candidate transaction}` (because a parent must have been committable). It is also easy to get it from the transaction placement (i.e. `placement.parent_tx_root`).

Hence, all we need is 2 `HashMap`s here instead of one :)

Reply:
Nice. Thank you!